Домашнее задание 4 урока

Prokopyevt opened the pull request:
1. Добавлен Lombok
2. Добавлен mapstruct
3. Добавлен swagger

Commit:
Домашнее задание 4 урока:
1. Добавлен Lombok
2. Добавлен mapstruct
3. Добавлен swagger

Korgutlova commented on src/main/java/com/simbirsoft/springcourse/service/Impl/LibraryServiceImpl.java:
```diff
@@ -1,24 +1,25 @@
 public class LibraryServiceImpl implements LibraryService {
 
-
+    private LibraryMapper libraryMapper;
```
final модификатор  (для всех остальных тоже, где еще встречается)

Korgutlova commented on src/main/java/com/simbirsoft/springcourse/service/Impl/ReaderServiceImpl.java:
```diff
@@ -32,13 +31,10 @@ public Reader getById(Long id) {
-        reader.setDateOfBirth(readerDto.getDateOfBirth());
-        return readerRepository.save(reader);
+      //  if (isEmpty(readerDto)) {
+        //    throw new NullPointerException("Пустое значение");
```
удаляем неиспользуемый код

Korgutlova commented on pom.xml:
```diff
@@ -24,6 +24,22 @@
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct</artifactId>
+			<version>1.4.1.Final</version>
```
значения версий обычно выносят в пропертис , где у тебя указана версия java

Korgutlova reviewed (changes requested):
В задании про swagger было сказано : "Добавить описания полей, методов, ответов от эндпоинта, добавить примеры ответов аннотациями" 

У тебя сделано описание методов с помощью `@ApiParam` правда это используется обычно для входных параметров, а для описания метода используется `@ApiOperation`  так же еще можно ипользовать `@ApiResponse` или `@ApiResponses` для описаний кодов ответа. 

Пожалуйста добавь эти изменения также.

Commit:
Переделанное домашнее задание 4 урока:
1. Добавлены модификаторы final в необходимых местах
2. Вынесено значение версии MapStruck в Properties
3. Добавлено описание полей, методов, ответов от эндпоинта.
Вопрос - считаются ли ответы кодов ошибок -  примерами ответов аннотациями ?
У меня не получилось найти примеры,не могли бы вы дать подсказку, если что-то не так. Заранее спасибо

Korgutlova reviewed (approved):
В принципе можно было для `@ApiParam` которые ты указал в модели добавить и example.

Пример `@ApiParam(required = true, value = "Имя читателя", example="Вася")`

И вообще насколько мне известно `@ApiParam`  используется непосредственно для методов и входных параметров, а другая аннотация `@ApiModelProperty` для моделей как раз 
https://www.baeldung.com/swagger-apiparam-vs-apimodelproperty

Commit:
Переделанное домашнее задание 4 урока:
1. Добавил аннотацию @ApiModelProperty  и параметр example

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/Impl/LibraryServiceImpl.java:
```diff
@@ -33,15 +34,10 @@ public Library getById(Long id) {
-        library.setStatus(libraryDto.getStatus());
-        return libraryRepository.save(library);
+       if (isEmpty(libraryDto)) {
+           throw new NullPointerException("Пустое значение");
```
эта проверка будет автоматически сгенерена внутри метода маппера. Но если ты хочешь тут просто вывести свою ошибку с текстом, то ок

Prokopyevt replied:
Могу ли я замержить эту ветку ?

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/controller/ReaderController.java:
```diff
@@ -12,27 +17,39 @@
     }
 
+    @ApiOperation(value = "Введите дату рождения и имя читателя")
+    @ApiResponses(value = {@ApiResponse(code = 200, message = "Запрос успешно выполнен"),
```
тут можно в описание кода 201 добавить модель ответа @Schema

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/controller/ReaderController.java:
```diff
@@ -12,27 +17,39 @@
-    }
-
+    @ApiOperation(value = "Введите необходимый ID")
+    @ApiResponses(value = {@ApiResponse(code = 200, message = "Запрос успешно выполнен"),
```
тут можно в описание кода 200 добавить модель ответа @Schema
Но это не критично, просто можешь потом сам добавить, если захочешь. Если добавишь @Schema, то в сваггере будет видна структура шаблонного ответа, то есть класса Reader.

ramilmmk reviewed (approved):
В целом ок, посмотри мой коммент насчёт добавления @Schema в ответы. Аннотация @Schema внутри @ApiResponse описывает примерную структуру ответа от сервера. То есть в интерфейсе сваггера будет видна структура этих классов. Это не критичное замечание, можешь смёржить сейчас, если не хочешь это добавлять.
